synom/src/cursor.rs

```diff
@@ -0,0 +1,357 @@
+//! This module defines a cheaply-copyable cursor into a TokenStream's data.
+//!
+//! It does this by copying the data into a stably-addressed structured buffer,
+//! and holding raw pointers into that buffer to allow walking through delimited
+//! sequences cheaply.
+//!
+//! This module is heavily commented as it contains the only unsafe code in
+//! `syn`, and caution should be made when editing it. It provides a safe
+//! interface, but is fragile internally.
+
+use proc_macro2::*;
+
+use std::ptr;
+use std::fmt;
+use std::marker::PhantomData;
+
+/// Internal type which is used instead of `TokenTree` to represent a single
+/// `TokenTree` within a `SynomBuffer`.
+#[derive(Debug)]
+enum Entry {
+    /// Mimicing types from proc-macro
+    Sequence(Span, Delimiter, SynomBuffer),
+    Word(Span, Symbol),
+    Op(Span, char, OpKind),
+    Literal(Span, Literal),
+    /// End entries contain a raw pointer to the entry from the containing TokenTree
+    End(*const Entry),
+}
+
+// A sync impl is necessary here because we need a global Entry object for the
+// empty sequence case. This impl is requied as `*const Entry` is neither `Send`
+// nor `Sync` by default.
+unsafe impl Sync for Entry {}
```

Comment:
I'm fixing this in #157. We can't make `Entry` Sync because that would make `SynomBuffer` sync implicitly, which we definitely don't want. I added a custom wrapper type instead.